diff --git a/libavcodec/Makefile b/libavcodec/Makefile
index 16d60b2..f922ae1 100644
--- a/libavcodec/Makefile
+++ b/libavcodec/Makefile
@@ -1255,6 +1255,7 @@ OBJS-$(CONFIG_APV_PARSER)              += apv_parser.o
 OBJS-$(CONFIG_AV1_PARSER)              += av1_parser.o av1_parse.o
 OBJS-$(CONFIG_AVS2_PARSER)             += avs2.o avs2_parser.o
 OBJS-$(CONFIG_AVS3_PARSER)             += avs3_parser.o
+OBJS-$(CONFIG_AV3A_PARSER)             += av3a.o av3a_parser.o
 OBJS-$(CONFIG_BMP_PARSER)              += bmp_parser.o
 OBJS-$(CONFIG_CAVSVIDEO_PARSER)        += cavs_parser.o
 OBJS-$(CONFIG_COOK_PARSER)             += cook_parser.o
diff --git a/libavcodec/av3a.c b/libavcodec/av3a.c
new file mode 100644
index 0000000..8f92ebc
--- /dev/null
+++ b/libavcodec/av3a.c
@@ -0,0 +1,364 @@
+/*
+ * Audio Vivid Codec
+ *
+ * This file is part of FFmpeg.
+ *
+ * FFmpeg is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * FFmpeg is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with FFmpeg; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+#include <stdio.h>
+#include "get_bits.h"
+#include "libavutil/channel_layout.h"
+#include "libavutil/samplefmt.h"
+#include "av3a.h"
+
+const McChanelConfig mcChannelConfigTable[AVS3_SIZE_MC_CONFIG_TABLE] = {
+    {"STEREO", CHANNEL_CONFIG_STEREO, 2},
+    {"MC_5_1_0", CHANNEL_CONFIG_MC_5_1, 6},
+    {"MC_7_1_0", CHANNEL_CONFIG_MC_7_1, 8},
+    {"MC_10_2", CHANNEL_CONFIG_MC_10_2, 12},
+    {"MC_22_2", CHANNEL_CONFIG_MC_22_2, 24},
+    {"MC_4_0", CHANNEL_CONFIG_MC_4_0, 4},
+    {"MC_5_1_2", CHANNEL_CONFIG_MC_5_1_2, 8},
+    {"MC_5_1_4", CHANNEL_CONFIG_MC_5_1_4, 10},
+    {"MC_7_1_2", CHANNEL_CONFIG_MC_7_1_2, 10},
+    {"MC_7_1_4", CHANNEL_CONFIG_MC_7_1_4, 12}};
+
+const int avs3_samplingrate_table[AVS3_SIZE_FS_TABLE] = {
+    192000, 96000, 48000, 44100, 32000, 24000, 22050, 16000, 8000};
+
+// bitrate table for mono
+const int64_t bitrateTableMono[AVS3_SIZE_BITRATE_TABLE] = {
+    16000, 32000, 44000, 56000, 64000, 72000, 80000, 96000, 128000, 144000,
+    164000, 192000, 0, 0, 0, 0};
+
+// bitrate table for stereo
+const int64_t bitrateTableStereo[AVS3_SIZE_BITRATE_TABLE] = {
+    24000, 32000, 48000, 64000, 80000, 96000, 128000, 144000, 192000, 256000,
+    320000, 0, 0, 0, 0, 0};
+
+// bitrate table for MC 5.1
+const int64_t bitrateTableMC5P1[AVS3_SIZE_BITRATE_TABLE] = {
+    192000, 256000, 320000, 384000, 448000, 512000, 640000, 720000, 144000, 96000,
+    128000, 160000, 0, 0, 0, 0};
+
+// bitrate table for MC 7.1
+const int64_t bitrateTableMC7P1[AVS3_SIZE_BITRATE_TABLE] = {
+    192000, 480000, 256000, 384000, 576000, 640000, 128000, 160000, 0, 0,
+    0, 0, 0, 0, 0, 0};
+
+// bitrate table for MC 4.0
+const int64_t bitrateTableMC4P0[AVS3_SIZE_BITRATE_TABLE] = {
+    48000, 96000, 128000, 192000, 256000, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0};
+
+// bitrate table for MC 5.1.2
+const int64_t bitrateTableMC5P1P2[AVS3_SIZE_BITRATE_TABLE] = {
+    152000, 320000, 480000, 576000, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0};
+
+// bitrate table for MC 5.1.4
+const int64_t bitrateTableMC5P1P4[AVS3_SIZE_BITRATE_TABLE] = {
+    176000, 384000, 576000, 704000, 256000, 448000, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0};
+
+// bitrate table for MC 7.1.2
+const int64_t bitrateTableMC7P1P2[AVS3_SIZE_BITRATE_TABLE] = {
+    216000, 480000, 576000, 384000, 768000, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0};
+
+// bitrate table for MC 7.1.4
+const int64_t bitrateTableMC7P1P4[AVS3_SIZE_BITRATE_TABLE] = {
+    240000, 608000, 384000, 512000, 832000, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0};
+
+const int64_t bitrateTableFoa[AVS3_SIZE_BITRATE_TABLE] = {
+    48000, 96000, 128000, 192000, 256000, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0};
+
+const int64_t bitrateTableHoa2[AVS3_SIZE_BITRATE_TABLE] = {
+    192000, 256000, 320000, 384000, 480000, 512000, 640000, 0, 0, 0,
+    0, 0, 0, 0, 0, 0};
+
+// bitrate table for HOA order 3
+const int64_t bitrateTableHoa3[AVS3_SIZE_BITRATE_TABLE] = {
+    256000, 320000, 384000, 512000, 640000, 896000, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0};
+
+// Codec channel number & bitrate config table
+// format: {channelConfigIdx, numChannels, bitrateTable}
+const CodecBitrateConfig codecBitrateConfigTable[CHANNEL_CONFIG_UNKNOWN] = {
+    {CHANNEL_CONFIG_MONO, bitrateTableMono},
+    {CHANNEL_CONFIG_STEREO, bitrateTableStereo},
+    {CHANNEL_CONFIG_MC_5_1, bitrateTableMC5P1},
+    {CHANNEL_CONFIG_MC_7_1, bitrateTableMC7P1},
+    {CHANNEL_CONFIG_MC_10_2, NULL},
+    {CHANNEL_CONFIG_MC_22_2, NULL},
+    {CHANNEL_CONFIG_MC_4_0, bitrateTableMC4P0},
+    {CHANNEL_CONFIG_MC_5_1_2, bitrateTableMC5P1P2},
+    {CHANNEL_CONFIG_MC_5_1_4, bitrateTableMC5P1P4},
+    {CHANNEL_CONFIG_MC_7_1_2, bitrateTableMC7P1P2},
+    {CHANNEL_CONFIG_MC_7_1_4, bitrateTableMC7P1P4},
+    {CHANNEL_CONFIG_HOA_ORDER1, bitrateTableFoa},
+    {CHANNEL_CONFIG_HOA_ORDER2, bitrateTableHoa2},
+    {CHANNEL_CONFIG_HOA_ORDER3, bitrateTableHoa3}};
+
+uint16_t read_av3a_frame_header(AVS3AHeaderInfo *hdf, const uint8_t *buf, const int byte_size){
+
+    GetBitContext gb;
+
+    uint8_t nn_type, codec_id, content_type, samping_rate_index, hoa_order, resolution_index, bitdepth, resolution;
+    AVS3AChannelConfig channel_config;
+
+    int16_t channels, objects;
+    uint64_t channel_layout;
+    int64_t bitratePerObj, bitrateBedMc, total_bitrate;
+    uint8_t coding_profile, num_chan_index, obj_brt_idx, bed_brt_idx, brt_idx;
+
+    // Read max header length into bs buffer
+    init_get_bits8(&gb, buf, MAX_NBYTES_FRAME_HEADER);
+
+    // 12 bits for frame sync word
+    if (get_bits(&gb, 12) != AVS3_AUDIO_SYNC_WORD){
+        return AVERROR_INVALIDDATA;
+    }
+
+    // 4 bits for codec id
+    codec_id = get_bits(&gb, 4);
+    if (codec_id != 2){
+        return AVERROR_INVALIDDATA;
+    }
+
+    // 1 bits for anc data
+    if (get_bits(&gb, 1) != 0){
+        return AVERROR_INVALIDDATA;
+    }
+
+    // 3 bits nn type
+    nn_type = get_bits(&gb, 3); 
+
+    // 3 bits for coding profile
+    coding_profile = get_bits(&gb, 3);
+
+    // 4 bits for sampling index
+    samping_rate_index = get_bits(&gb, 4);
+
+    // skip 8 bits for CRC first part
+    skip_bits(&gb, 8);
+
+    if (coding_profile == 0){
+
+        content_type = 0;
+
+        // 7 bits for mono/stereo/MC
+        num_chan_index = get_bits(&gb, 7);
+
+        channel_config = (AVS3AChannelConfig)num_chan_index;
+        
+        switch (channel_config)
+        {
+        case CHANNEL_CONFIG_MONO:
+            channels = 1;
+            channel_layout = AV_CH_LAYOUT_MONO;
+            break;
+        case CHANNEL_CONFIG_STEREO:
+            channels = 2;
+            channel_layout = AV_CH_LAYOUT_STEREO;
+            break;
+        case CHANNEL_CONFIG_MC_4_0:
+            channels = 4;
+            break;
+        case CHANNEL_CONFIG_MC_5_1:
+            channels = 6;
+            channel_layout = AVS3P3_CH_LAYOUT_5POINT1;
+            break;
+        case CHANNEL_CONFIG_MC_7_1:
+            channels = 8;
+            channel_layout = AV_CH_LAYOUT_7POINT1;
+            break;
+        case CHANNEL_CONFIG_MC_5_1_2:
+            channels = 8;
+            break;
+        case CHANNEL_CONFIG_MC_5_1_4:
+            channels = 10;
+            break;
+        case CHANNEL_CONFIG_MC_7_1_2:
+            channels = 10;
+            break;
+        case CHANNEL_CONFIG_MC_7_1_4:
+            channels = 12;
+            break;
+        case CHANNEL_CONFIG_MC_22_2:
+            channels = 24;
+            channel_layout = AV_CH_LAYOUT_22POINT2;
+            break;
+        default:
+            break;
+        }
+    }
+    else if (coding_profile == 1){
+
+        // sound bed type, 2bits
+        uint8_t soundBedType = get_bits(&gb, 2);
+
+        if (soundBedType == 0){   
+
+            content_type = 1;
+
+            // for only objects
+            // object number, 7 bits
+            objects = get_bits(&gb, 7);
+            objects += 1;
+
+            // bitrate index for each obj, 4 bits
+            obj_brt_idx = get_bits(&gb, 4);
+
+            total_bitrate = codecBitrateConfigTable[CHANNEL_CONFIG_MONO].bitrateTable[obj_brt_idx] * objects;
+        }
+        else if (soundBedType == 1){
+
+            content_type = 2;
+
+            // for MC + objs
+            // channel number index, 7 bits
+            num_chan_index = get_bits(&gb, 7);
+
+            // bitrate index for sound bed, 4 bits
+            bed_brt_idx = get_bits(&gb, 4);
+
+            // object number, 7 bits
+            objects = get_bits(&gb, 7);
+            objects += 1;
+
+            // bitrate index for each obj, 4 bits
+            obj_brt_idx = get_bits(&gb, 4);
+
+            // channelNumIdx for sound bed
+            channel_config = (AVS3AChannelConfig)num_chan_index;
+
+            // sound bed bitrate
+            bitrateBedMc = codecBitrateConfigTable[channel_config].bitrateTable[bed_brt_idx];
+
+            // numChannels for sound bed
+            for (int16_t i = 0; i < AVS3_SIZE_MC_CONFIG_TABLE; i++)
+            {
+                if (channel_config == mcChannelConfigTable[i].channelNumConfig)
+                {
+                    channels = mcChannelConfigTable[i].numChannels;
+                }
+            }
+
+            // bitrate per obj
+            bitratePerObj = codecBitrateConfigTable[CHANNEL_CONFIG_MONO].bitrateTable[obj_brt_idx];
+
+            // add num chans and num objs to get total chans
+            /* channels += objects; */
+
+            total_bitrate = bitrateBedMc + bitratePerObj * objects;
+        }
+    }
+    else if (coding_profile == 2){
+        content_type = 3;
+        
+        // 4 bits for HOA order
+        hoa_order = get_bits(&gb, 4);
+        hoa_order += 1;
+
+        switch (hoa_order)
+        {
+        case 1:
+            channels = 4;
+            channel_config = CHANNEL_CONFIG_HOA_ORDER1;
+            break;
+        case 2:
+            channels = 9;
+            channel_config = CHANNEL_CONFIG_HOA_ORDER2;
+            break;
+        case 3:
+            channels = 16;
+            channel_config = CHANNEL_CONFIG_HOA_ORDER3;
+            break;
+        default:
+            break;
+        }
+    }else{
+        return AVERROR_INVALIDDATA;
+    }
+
+    // 2 bits for bit depth
+    resolution_index = get_bits(&gb, 2);
+    switch (resolution_index){
+    case 0:
+        bitdepth = AV_SAMPLE_FMT_U8;
+        resolution = 8;
+        break;
+    case 1:
+        bitdepth = AV_SAMPLE_FMT_S16;
+        resolution = 16;
+        break;
+    case 2:
+        resolution = 24;
+        break;
+    default:
+        return AVERROR_INVALIDDATA;
+    }
+
+    if (coding_profile != 1){
+        // 4 bits for bitrate index
+        brt_idx = get_bits(&gb, 4);
+        total_bitrate = codecBitrateConfigTable[channel_config].bitrateTable[brt_idx];
+    }
+
+    // 8 bits for CRC second part
+    skip_bits(&gb, 8);
+
+    /* AVS3P6 M6954-v3 */
+    hdf->codec_id = codec_id;
+    hdf->sampling_rate_index = samping_rate_index;
+    hdf->sampling_rate = avs3_samplingrate_table[samping_rate_index];
+    hdf->bitdepth = bitdepth;
+
+    hdf->nn_type = nn_type;
+    hdf->content_type = content_type;
+
+    if(hdf->content_type == 0){
+        hdf->channel_num_index = num_chan_index;
+        hdf->channels = channels;
+        hdf->objects = 0;
+        hdf->total_channels = channels; 
+        hdf->channel_layout = channel_layout;
+    }else if(hdf->content_type == 1){
+        hdf->objects = objects;
+        hdf->channels = 0;
+        hdf->total_channels = objects; 
+    }else if(hdf->content_type == 2){
+        hdf->channel_num_index = num_chan_index;
+        hdf->channels = channels;
+        hdf->objects = objects;
+        hdf->total_channels = channels + objects; 
+        hdf->channel_layout = channel_layout;
+    }else if(hdf->content_type == 3){
+        hdf->hoa_order = hoa_order;
+        hdf->channels = channels;
+        hdf->total_channels = channels; 
+    }else{
+        return AVERROR_INVALIDDATA;
+    }
+
+    hdf->total_bitrate = total_bitrate;
+    hdf->resolution = resolution;
+    hdf->resolution_index = resolution_index;
+
+    return 0;
+}
\ No newline at end of file
diff --git a/libavcodec/av3a.h b/libavcodec/av3a.h
new file mode 100644
index 0000000..0f5a8ae
--- /dev/null
+++ b/libavcodec/av3a.h
@@ -0,0 +1,124 @@
+/*
+ * Audio Vivid Codec
+ *
+ * This file is part of FFmpeg.
+ *
+ * FFmpeg is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * FFmpeg is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with FFmpeg; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+#include <stdint.h>
+#include "libavutil/channel_layout.h"
+
+/* AVS3 header */
+#define AVS3_AUDIO_HEADER_SIZE 7
+#define AVS3_SYNC_WORD_SIZE 2
+#define MAX_NBYTES_FRAME_HEADER 9
+#define AVS3_AUDIO_SYNC_WORD 0xFFF
+
+#define AVS3_AUDIO_FRAME_SIZE 1024
+#define AVS3_SIZE_BITRATE_TABLE 16
+#define AVS3_SIZE_FS_TABLE 9
+
+/* AVS3 Audio Format */
+#define AVS3_MONO_FORMAT 0
+#define AVS3_STEREO_FORMAT 1
+#define AVS3_MC_FORMAT 2
+#define AVS3_HOA_FORMAT 3
+#define AVS3_MIX_FORMAT 4
+
+#define AVS3_SIZE_MC_CONFIG_TABLE 10
+
+#define AVS3P3_CH_LAYOUT_5POINT1 (AV_CH_LAYOUT_SURROUND | AV_CH_LOW_FREQUENCY | AV_CH_BACK_LEFT | AV_CH_BACK_RIGHT)
+
+/* Sampling rate index table */
+extern const int avs3_samplingrate_table[AVS3_SIZE_FS_TABLE];
+
+// AVS3P3 header information
+typedef struct
+{
+    // header info
+    uint8_t codec_id;
+    uint8_t sampling_rate_index;
+    int sampling_rate;
+
+    uint16_t bitdepth;
+    uint16_t channels;
+    uint16_t objects;
+    uint16_t hoa_order;
+    uint64_t channel_layout;
+    int64_t total_bitrate;
+
+    // configuration
+    uint8_t content_type;
+    uint16_t channel_num_index;
+    uint16_t total_channels;
+    uint8_t resolution;
+    uint8_t nn_type;
+    uint8_t resolution_index;
+
+} AVS3AHeaderInfo;
+
+typedef enum
+{
+    CHANNEL_CONFIG_MONO = 0,
+    CHANNEL_CONFIG_STEREO = 1,
+    CHANNEL_CONFIG_MC_5_1,
+    CHANNEL_CONFIG_MC_7_1,
+    CHANNEL_CONFIG_MC_10_2,
+    CHANNEL_CONFIG_MC_22_2,
+    CHANNEL_CONFIG_MC_4_0,
+    CHANNEL_CONFIG_MC_5_1_2,
+    CHANNEL_CONFIG_MC_5_1_4,
+    CHANNEL_CONFIG_MC_7_1_2,
+    CHANNEL_CONFIG_MC_7_1_4,
+    CHANNEL_CONFIG_HOA_ORDER1,
+    CHANNEL_CONFIG_HOA_ORDER2,
+    CHANNEL_CONFIG_HOA_ORDER3,
+    CHANNEL_CONFIG_UNKNOWN
+} AVS3AChannelConfig;
+
+/* Codec bitrate config struct */
+typedef struct CodecBitrateConfigStructure
+{
+    AVS3AChannelConfig channelNumConfig;
+    const int64_t *bitrateTable;
+} CodecBitrateConfig;
+
+typedef struct McChannelConfigStructure
+{
+    const char mcCmdString[10];
+    AVS3AChannelConfig channelNumConfig;
+    const int16_t numChannels;
+} McChanelConfig;
+
+extern const McChanelConfig mcChannelConfigTable[AVS3_SIZE_MC_CONFIG_TABLE];
+
+/* Codec bitrate config struct */
+extern const CodecBitrateConfig codecBitrateConfigTable[CHANNEL_CONFIG_UNKNOWN];
+
+/* bitrate table for each mode */
+extern const int64_t bitrateTableMono[AVS3_SIZE_BITRATE_TABLE];
+extern const int64_t bitrateTableStereo[AVS3_SIZE_BITRATE_TABLE];
+extern const int64_t bitrateTableMC5P1[AVS3_SIZE_BITRATE_TABLE];
+extern const int64_t bitrateTableMC7P1[AVS3_SIZE_BITRATE_TABLE];
+extern const int64_t bitrateTableMC4P0[AVS3_SIZE_BITRATE_TABLE];
+extern const int64_t bitrateTableMC5P1P2[AVS3_SIZE_BITRATE_TABLE];
+extern const int64_t bitrateTableMC5P1P4[AVS3_SIZE_BITRATE_TABLE];
+extern const int64_t bitrateTableMC7P1P2[AVS3_SIZE_BITRATE_TABLE];
+extern const int64_t bitrateTableMC7P1P4[AVS3_SIZE_BITRATE_TABLE];
+extern const int64_t bitrateTableFoa[AVS3_SIZE_BITRATE_TABLE];
+extern const int64_t bitrateTableHoa2[AVS3_SIZE_BITRATE_TABLE];
+extern const int64_t bitrateTableHoa3[AVS3_SIZE_BITRATE_TABLE];
+
+extern uint16_t read_av3a_frame_header(AVS3AHeaderInfo *hdf, const uint8_t *buf, const int byte_size);
diff --git a/libavcodec/av3a_parser.c b/libavcodec/av3a_parser.c
new file mode 100644
index 0000000..ad028f2
--- /dev/null
+++ b/libavcodec/av3a_parser.c
@@ -0,0 +1,81 @@
+/*
+ * Audio Vivid Codec
+ *
+ * This file is part of FFmpeg.
+ *
+ * FFmpeg is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * FFmpeg is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with FFmpeg; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+
+#include "libavutil/samplefmt.h"
+#include "libavutil/intreadwrite.h"
+#include "libavutil/channel_layout.h"
+#include "parser.h"
+#include "parser_internal.h"
+#include "get_bits.h"
+#include "av3a.h"
+
+typedef struct AVS3AudioParseContext
+{
+    int frame_size;
+    int bitdepth;
+    int sample_rate;
+    uint64_t bit_rate;
+    uint16_t channels;
+    uint64_t channel_layout;
+
+} AVS3AParseContext;
+
+static int raw_av3a_parse(AVCodecParserContext *s, AVCodecContext *avctx,
+                          const uint8_t **poutbuf, int *poutbuf_size,
+                          const uint8_t *buf, int buf_size)
+{
+    int ret = 0;
+    uint8_t header[MAX_NBYTES_FRAME_HEADER];
+    AVS3AHeaderInfo hdf;
+
+    if (buf_size < MAX_NBYTES_FRAME_HEADER){
+        return buf_size;
+    }
+
+    memcpy(header, buf, MAX_NBYTES_FRAME_HEADER);
+
+    // read frame header
+    if ((ret = read_av3a_frame_header(&hdf, header, MAX_NBYTES_FRAME_HEADER)) != 0){
+        return ret;
+    }
+
+    avctx->sample_rate = hdf.sampling_rate;
+    avctx->bit_rate = hdf.total_bitrate;
+    av_channel_layout_uninit(&avctx->ch_layout);
+    if (hdf.channel_layout)
+        av_channel_layout_from_mask(&avctx->ch_layout, hdf.channel_layout);
+    else
+        avctx->ch_layout.order = AV_CHANNEL_ORDER_UNSPEC;
+    avctx->ch_layout.nb_channels = hdf.total_channels;
+    avctx->frame_size = AVS3_AUDIO_FRAME_SIZE;
+    s->format = hdf.bitdepth;
+
+    /* return the full packet */
+    *poutbuf = buf;
+    *poutbuf_size = buf_size;
+
+    return buf_size;
+}
+
+const FFCodecParser ff_av3a_parser = {
+    PARSER_CODEC_LIST(AV_CODEC_ID_AVS3_AUDIO),
+    .priv_data_size = sizeof(AVS3AParseContext),
+    .parse          = raw_av3a_parse,
+};
diff --git a/libavcodec/codec_desc.c b/libavcodec/codec_desc.c
index f5f2b17..f48361e 100644
--- a/libavcodec/codec_desc.c
+++ b/libavcodec/codec_desc.c
@@ -3564,6 +3564,13 @@ static const AVCodecDescriptor codec_descriptors[] = {
         .long_name = NULL_IF_CONFIG_SMALL("Apple Positional Audio Codec"),
         .props     = AV_CODEC_PROP_LOSSY,
     },
+    {
+        .id        = AV_CODEC_ID_AVS3_AUDIO,
+        .type      = AVMEDIA_TYPE_AUDIO,
+        .name      = "avs3_audio",
+        .long_name = NULL_IF_CONFIG_SMALL("AVS3-P3 3D Audio / Audio Vivid"),
+        .props     = AV_CODEC_PROP_LOSSY,
+    },
 
     /* subtitle codecs */
     {
diff --git a/libavcodec/codec_id.h b/libavcodec/codec_id.h
index 7cafac6..16e8039 100644
--- a/libavcodec/codec_id.h
+++ b/libavcodec/codec_id.h
@@ -561,6 +561,8 @@ enum AVCodecID {
     AV_CODEC_ID_AHX,
     AV_CODEC_ID_APPLE_APAC,
 
+    AV_CODEC_ID_AVS3_AUDIO,        ///< AVS3-P3 3D Audio / Audio Vivid
+
     /* subtitle codecs */
     AV_CODEC_ID_FIRST_SUBTITLE = 0x17000,          ///< A dummy ID pointing at the start of subtitle codecs.
     AV_CODEC_ID_DVD_SUBTITLE = 0x17000,
diff --git a/libavcodec/parsers.c b/libavcodec/parsers.c
index b54442a..c069b05 100644
--- a/libavcodec/parsers.c
+++ b/libavcodec/parsers.c
@@ -31,6 +31,7 @@ extern const FFCodecParser ff_apv_parser;
 extern const FFCodecParser ff_av1_parser;
 extern const FFCodecParser ff_avs2_parser;
 extern const FFCodecParser ff_avs3_parser;
+extern const FFCodecParser ff_av3a_parser;
 extern const FFCodecParser ff_bmp_parser;
 extern const FFCodecParser ff_cavsvideo_parser;
 extern const FFCodecParser ff_cook_parser;
diff --git a/libavformat/Makefile b/libavformat/Makefile
index 466f5d1..dd82573 100644
--- a/libavformat/Makefile
+++ b/libavformat/Makefile
@@ -158,6 +158,8 @@ OBJS-$(CONFIG_AVS2_DEMUXER)              += avs2dec.o rawdec.o
 OBJS-$(CONFIG_AVS2_MUXER)                += rawenc.o
 OBJS-$(CONFIG_AVS3_DEMUXER)              += avs3dec.o rawdec.o
 OBJS-$(CONFIG_AVS3_MUXER)                += rawenc.o
+OBJS-$(CONFIG_AV3A_DEMUXER)              += av3adec.o
+OBJS-$(CONFIG_AV3A_MUXER)                += rawenc.o
 OBJS-$(CONFIG_BETHSOFTVID_DEMUXER)       += bethsoftvid.o
 OBJS-$(CONFIG_BFI_DEMUXER)               += bfi.o
 OBJS-$(CONFIG_BINK_DEMUXER)              += bink.o
diff --git a/libavformat/allformats.c b/libavformat/allformats.c
index e121c74..d04ff2f 100644
--- a/libavformat/allformats.c
+++ b/libavformat/allformats.c
@@ -101,6 +101,8 @@ extern const FFInputFormat  ff_avs2_demuxer;
 extern const FFOutputFormat ff_avs2_muxer;
 extern const FFInputFormat  ff_avs3_demuxer;
 extern const FFOutputFormat ff_avs3_muxer;
+extern const FFInputFormat  ff_av3a_demuxer;
+extern const FFOutputFormat ff_av3a_muxer;
 extern const FFInputFormat  ff_bethsoftvid_demuxer;
 extern const FFInputFormat  ff_bfi_demuxer;
 extern const FFInputFormat  ff_bintext_demuxer;
diff --git a/libavformat/av3adec.c b/libavformat/av3adec.c
new file mode 100644
index 0000000..056eabd
--- /dev/null
+++ b/libavformat/av3adec.c
@@ -0,0 +1,397 @@
+/*
+ * Audio Vivid Codec
+ *
+ * This file is part of FFmpeg.
+ *
+ * FFmpeg is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * FFmpeg is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with FFmpeg; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+#include "demux.h"
+#include "avio_internal.h"
+#include "internal.h"
+#include "apetag.h"
+#include "rawdec.h"
+#include "libavutil/opt.h"
+#include "libavutil/avassert.h"
+#include "libavutil/intreadwrite.h"
+#include "libavutil/channel_layout.h"
+#include "libavutil/mem.h"
+#include "libavcodec/get_bits.h"
+#include "libavcodec/av3a.h"
+#include <string.h>
+#include <math.h>
+
+typedef struct
+{
+    uint64_t file_size;
+    uint64_t frame_cnt;
+} AV3ADecContext;
+
+/* Bytes assignment */
+typedef struct
+{
+    uint8_t audio_codec_id;
+    uint8_t sampling_rate_index;
+    uint8_t nn_type;
+    uint8_t content_type;
+    uint8_t channel_num_index;
+    uint8_t num_objects;
+    uint8_t hoa_order;
+    uint8_t resolution_index;
+    uint16_t total_bitrate_kbps;
+} AV3ASpecificContext;
+
+static int get_av3a_payload(AVFormatContext *s)
+{
+
+    int read_bytes = 0;
+    uint16_t sync_word = 0;
+    uint8_t header[MAX_NBYTES_FRAME_HEADER];
+    GetBitContext gb;
+    uint8_t samping_index, num_channels_idx, bitrate_index, soundbed_type, hoa_order;
+    uint8_t coding_profile;
+    int sample_rate;
+    uint64_t bit_rate;
+    AVS3AChannelConfig channel_config;
+    int payload_bytes = 0;
+    uint16_t bitrateIdxPerObj, bitrateIdxBedMc;
+    int numObjs, numChans;
+
+    if ((read_bytes = avio_read(s->pb, header, MAX_NBYTES_FRAME_HEADER)) != MAX_NBYTES_FRAME_HEADER)
+    {
+
+        if (avio_feof(s->pb))
+        {
+            return AVERROR_EOF;
+        }
+        return AVERROR_INVALIDDATA;
+    }
+
+    init_get_bits8(&gb, header, MAX_NBYTES_FRAME_HEADER);
+
+    // 12 bits for frame sync word
+    sync_word = get_bits(&gb, 12);
+
+    if (sync_word != AVS3_AUDIO_SYNC_WORD)
+    {
+        return AVERROR_INVALIDDATA;
+    }
+
+    // skip audio_codec_id 4 bits & anc data 1 bit & 3 bits nn type
+    skip_bits(&gb, 5 + 3);
+
+    coding_profile = get_bits(&gb, 3);
+
+    samping_index = get_bits(&gb, 4);
+
+    // skip crc first
+    skip_bits(&gb, 8);
+
+    if (coding_profile == 0)
+    {
+
+        num_channels_idx = get_bits(&gb, 7);
+        channel_config = (AVS3AChannelConfig)num_channels_idx;
+    }
+    else if (coding_profile == 1)
+    {
+
+        soundbed_type = get_bits(&gb, 2);
+
+        if (soundbed_type == 0)
+        {
+            numObjs = get_bits(&gb, 7);
+            numObjs += 1;
+            bitrateIdxPerObj = get_bits(&gb, 4);
+            bit_rate = codecBitrateConfigTable[CHANNEL_CONFIG_MONO].bitrateTable[bitrateIdxPerObj] * numObjs;
+        }
+        else if (soundbed_type == 1)
+        {
+
+            uint64_t bitratePerObj, bitrateBedMc;
+
+            num_channels_idx = get_bits(&gb, 7);
+            bitrateIdxBedMc = get_bits(&gb, 4);
+            numObjs = get_bits(&gb, 7);
+            numObjs += 1;
+            bitrateIdxPerObj = get_bits(&gb, 4);
+
+            channel_config = (AVS3AChannelConfig)num_channels_idx;
+
+            // numChannels for sound bed
+            for (int16_t i = 0; i < AVS3_SIZE_MC_CONFIG_TABLE; i++)
+            {
+                if (channel_config == mcChannelConfigTable[i].channelNumConfig)
+                {
+                    numChans = mcChannelConfigTable[i].numChannels;
+                }
+            }
+
+            bitrateBedMc = codecBitrateConfigTable[channel_config].bitrateTable[bitrateIdxBedMc];
+
+            // bitrate per obj
+            bitratePerObj = codecBitrateConfigTable[CHANNEL_CONFIG_MONO].bitrateTable[bitrateIdxPerObj];
+
+            // add num chans and num objs to get total chans
+            numChans += numObjs;
+
+            bit_rate = bitrateBedMc + bitratePerObj * numObjs;
+        }
+    }
+    else if (coding_profile == 2)
+    {
+
+        hoa_order = get_bits(&gb, 4);
+        hoa_order += 1;
+        switch (hoa_order)
+        {
+        case 1:
+            numChans = 4;
+            channel_config = CHANNEL_CONFIG_HOA_ORDER1;
+            break;
+        case 2:
+            numChans = 9;
+            channel_config = CHANNEL_CONFIG_HOA_ORDER2;
+            break;
+        case 3:
+            numChans = 16;
+            channel_config = CHANNEL_CONFIG_HOA_ORDER3;
+            break;
+        default:
+            break;
+        }
+    }
+
+    // skip bit depth 2 bits
+    skip_bits(&gb, 2);
+
+    if (coding_profile != 1)
+    {
+        bitrate_index = get_bits(&gb, 4);
+        bit_rate = codecBitrateConfigTable[channel_config].bitrateTable[bitrate_index];
+    }
+
+    // skip crc second part 8 bits
+    skip_bits(&gb, 8);
+
+    sample_rate = avs3_samplingrate_table[samping_index];
+
+    payload_bytes = (int)ceil((((float)(bit_rate) / sample_rate) * AVS3_AUDIO_FRAME_SIZE) / 8);
+
+    avio_seek(s->pb, -read_bytes, SEEK_CUR);
+
+    return payload_bytes;
+}
+
+static int av3a_probe(const AVProbeData *p)
+{
+
+    uint16_t frame_sync_word;
+    const char *ptr_ext = NULL;
+    uint16_t lval = ((uint16_t)(p->buf[0]));
+    uint16_t rval = ((uint16_t)(p->buf[1]));
+    char ext_str[5]; // .av3a
+    int len = 0;
+    int len_ext = 0;
+    frame_sync_word = ((lval << 8) | rval) >> 4;
+
+    /* check sync word */
+    if (frame_sync_word == AVS3_AUDIO_SYNC_WORD)
+    {
+        len_ext = strlen(".av3a");
+        av_assert0(len_ext == 5);
+        if (p->filename != NULL)
+        {
+            len = strlen(p->filename);
+            if (len >= len_ext)
+            {
+                memcpy(ext_str, p->filename + len - len_ext, len_ext);
+                ptr_ext = strrchr(ext_str, '.');
+            }
+        }
+
+        if (ptr_ext != NULL && strcmp(ptr_ext, ".av3a") == 0)
+        {
+            return AVPROBE_SCORE_MAX;
+        }
+    }
+
+    return 0;
+}
+
+static int av3a_read_header(AVFormatContext *s)
+{
+
+    uint8_t header[MAX_NBYTES_FRAME_HEADER];
+    AVIOContext *pb = s->pb; // set by avformat_open_input.
+    AVStream *stream = NULL;
+    AV3ADecContext *av3adecCtx = s->priv_data;
+    AVDictionary *extra_language_metada = NULL;
+    AVS3AHeaderInfo hdf;
+    int ret = 0;
+    AV3ASpecificContext av3aSpecificCtx;
+
+    // create input stream(AVStream) in AVFormatContext->streams = stream.
+    if (!(stream = avformat_new_stream(s, NULL)))
+    {
+        return AVERROR(ENOMEM);
+    }
+
+    // set stream constant field
+    stream->start_time = 0;
+    stream->codecpar->codec_type = AVMEDIA_TYPE_AUDIO;
+    stream->codecpar->codec_id = ffifmt(s->iformat)->raw_codec_id;   // id = AV_CODEC_ID_AVS3_AUDIO
+    stream->codecpar->codec_tag = MKTAG('a', 'v', '3', 'a'); // Tag = 'av3a'
+    ffstream(stream)->need_parsing = AVSTREAM_PARSE_FULL_RAW;
+
+    // set language metadata
+    av_dict_set(&extra_language_metada, "language", "eng", 0);
+    stream->metadata = extra_language_metada;
+
+    // read header with max header size
+    avio_read(pb, header, MAX_NBYTES_FRAME_HEADER);
+
+    // sparse frame header bits
+    if ((ret = read_av3a_frame_header(&hdf, header, MAX_NBYTES_FRAME_HEADER)) != 0)
+    {
+        return ret;
+    }
+
+    // codec parameters
+    stream->codecpar->format = hdf.bitdepth;
+    stream->codecpar->bit_rate = hdf.total_bitrate;
+    stream->codecpar->sample_rate = (int)(hdf.sampling_rate);
+    stream->codecpar->frame_size = AVS3_AUDIO_FRAME_SIZE; // fixed frame length
+    av_channel_layout_uninit(&stream->codecpar->ch_layout);
+    if (hdf.channel_layout)
+        av_channel_layout_from_mask(&stream->codecpar->ch_layout, hdf.channel_layout);
+    else
+        stream->codecpar->ch_layout.order = AV_CHANNEL_ORDER_UNSPEC;
+    stream->codecpar->ch_layout.nb_channels = hdf.total_channels;
+    stream->codecpar->bits_per_raw_sample = hdf.resolution;
+
+    // configuration parameters
+    av3aSpecificCtx.audio_codec_id = hdf.codec_id;
+    av3aSpecificCtx.sampling_rate_index = hdf.sampling_rate_index;
+    av3aSpecificCtx.nn_type = hdf.nn_type;
+    av3aSpecificCtx.content_type = hdf.content_type;
+    av3aSpecificCtx.channel_num_index = hdf.channel_num_index;
+    av3aSpecificCtx.num_objects = hdf.objects;
+    av3aSpecificCtx.hoa_order = hdf.hoa_order;
+    av3aSpecificCtx.resolution_index = hdf.resolution_index;
+    av3aSpecificCtx.total_bitrate_kbps = (int)(hdf.total_bitrate / 1000);
+
+    if (!stream->codecpar->extradata)
+    {
+        stream->codecpar->extradata = av_malloc(sizeof(AV3ASpecificContext) + AV_INPUT_BUFFER_PADDING_SIZE);
+        if (!stream->codecpar->extradata)
+        {
+            return AVERROR(ENOMEM);
+        }
+        stream->codecpar->extradata_size = sizeof(AV3ASpecificContext);
+        memcpy(stream->codecpar->extradata, &av3aSpecificCtx, sizeof(AV3ASpecificContext));
+    }
+
+    av3adecCtx->file_size = 0L;
+    av3adecCtx->frame_cnt = 0L;
+
+    // seek back to bitstream start offset.
+    avio_seek(s->pb, -MAX_NBYTES_FRAME_HEADER, SEEK_CUR);
+
+    return 0;
+}
+
+static int av3a_read_packet(AVFormatContext *s, AVPacket *pkt)
+{
+
+    int64_t pos;
+    int packet_size = 0;
+    int read_bytes = 0;
+    AV3ADecContext *av3adecCtx = s->priv_data;
+
+    if (!s->streams[0]->codecpar)
+    {
+        return AVERROR(ENOMEM);
+    }
+
+    if (avio_feof(s->pb))
+    {
+        return AVERROR_EOF;
+    }
+
+    pos = avio_tell(s->pb);
+
+    // get packet size
+    if ((packet_size = get_av3a_payload(s)) < 0)
+    {
+        return packet_size;
+    }
+
+    // create packet buffer to encoded frame(AVPacket is public ABI).
+    if (packet_size == 0 || av_new_packet(pkt, packet_size))
+    {
+        return AVERROR(EIO);
+    }
+
+    // just audio stream
+    pkt->stream_index = 0;
+    pkt->pos = pos;
+    pkt->duration = s->streams[0]->codecpar->frame_size;
+    // pkt->dts;
+
+    if ((read_bytes = avio_read(s->pb, pkt->data, packet_size)) < 0)
+    {
+        if (read_bytes >= 0 && read_bytes < packet_size)
+        {
+            pkt->size = read_bytes;
+        }
+        else
+        {
+
+            if (avio_feof(s->pb))
+            {
+                pkt->size = 0;
+                return AVERROR_EOF;
+            }
+            else
+            {
+                return AVERROR(EIO);
+            }
+        }
+    }
+
+    if (av3adecCtx->frame_cnt < UINT64_MAX - 1)
+    {
+        av3adecCtx->frame_cnt++;
+        if (av3adecCtx->file_size < UINT64_MAX - 1)
+        {
+            av3adecCtx->file_size += read_bytes;
+        }
+    }
+
+    return 0;
+}
+
+const FFInputFormat ff_av3a_demuxer = {
+    .p.name         = "av3a",
+    .p.long_name    = NULL_IF_CONFIG_SMALL("AVS3-P3 3D Audio/Audio Vivid AATF bitstream demuxer"),
+    .p.extensions   = "av3a",
+    .p.flags        = AVFMT_GENERIC_INDEX,
+    .p.mime_type    = "audio/av3a",
+    .raw_codec_id   = AV_CODEC_ID_AVS3_AUDIO,
+    .priv_data_size = sizeof(AV3ADecContext),
+    .read_probe     = av3a_probe,
+    .read_header    = av3a_read_header,
+    .read_packet    = av3a_read_packet,
+};
diff --git a/libavformat/isom_tags.c b/libavformat/isom_tags.c
index c0c239d..8ed87a5 100644
--- a/libavformat/isom_tags.c
+++ b/libavformat/isom_tags.c
@@ -309,6 +309,7 @@ const AVCodecTag ff_codec_movaudio_tags[] = {
     { AV_CODEC_ID_AC3,             MKTAG('a', 'c', '-', '3') }, /* ETSI TS 102 366 Annex F */
     { AV_CODEC_ID_AC3,             MKTAG('s', 'a', 'c', '3') }, /* Nero Recode */
     { AV_CODEC_ID_AC4,             MKTAG('a', 'c', '-', '4') },
+    { AV_CODEC_ID_AVS3_AUDIO,      MKTAG('a', 'v', '3', 'a') },
     { AV_CODEC_ID_ADPCM_IMA_QT,    MKTAG('i', 'm', 'a', '4') },
     { AV_CODEC_ID_ALAC,            MKTAG('a', 'l', 'a', 'c') },
     { AV_CODEC_ID_AMR_NB,          MKTAG('s', 'a', 'm', 'r') }, /* AMR-NB 3gp */
diff --git a/libavformat/mov.c b/libavformat/mov.c
index f53ce69..e63553f 100644
--- a/libavformat/mov.c
+++ b/libavformat/mov.c
@@ -1234,6 +1234,104 @@ static uint64_t mov_dts_channel_layout_to_mask(unsigned int code)
     return mask;
 }
 
+typedef struct {
+    const char*  tag;
+    uint8_t  channels;
+} ff_av3a_chan_map;
+
+static const ff_av3a_chan_map av3a_chan_config_table[] = {
+    {"Mono",   1},
+    {"Stereo", 2},
+    {"5.1",    6},
+    {"7.1",    8},
+    {"10.2",   12},
+    {"22.2",   24},
+    {"4.0",    4},
+    {"5.1.2",  8},
+    {"5.1.4",  10},
+    {"7.1.2",  10},
+    {"7.1.4",  12},
+    {"Unknown", 0}
+};
+
+static int mov_read_dca3(MOVContext *c, AVIOContext *pb, MOVAtom atom)
+{
+    AVStream *st;
+    uint8_t buff[7];
+    GetBitContext gb;
+    uint8_t  codec_id, sr_idx, nn_type, content_type, ch_idx, num_objects;
+    uint8_t  hoa_order;
+    uint16_t brt_kbps;
+    uint8_t  resolution, resolution_idx;
+    uint32_t sr_map[] = { 192000, 96000, 48000, 44100, 32000, 24000, 22050, 16000, 8000};
+    uint8_t res_map[] = { 8, 16, 24};
+    ff_av3a_chan_map chan_map;
+
+    if (c->fc->nb_streams < 1)
+        return 0;
+    st = c->fc->streams[c->fc->nb_streams-1];
+
+    avio_read(pb, buff, sizeof(buff));
+
+    init_get_bits8(&gb, buff, sizeof(buff));
+
+    codec_id = get_bits(&gb, 4);
+    sr_idx = get_bits(&gb, 4);
+
+    if(codec_id == 2){
+        nn_type      = get_bits(&gb, 3);
+        content_type = get_bits(&gb, 3);
+
+        if (content_type == 0){
+            ch_idx = get_bits(&gb, 8);
+        }
+        else if (content_type == 1){
+            num_objects = get_bits(&gb, 8);
+        }
+        else if (content_type == 2){
+            ch_idx = get_bits(&gb, 8);
+            num_objects = get_bits(&gb, 8);
+        }
+        else if (content_type == 3){
+            hoa_order = get_bits(&gb, 4);
+        }
+
+        brt_kbps = get_bits(&gb, 16);
+    }
+
+    resolution_idx = get_bits(&gb, 2);
+    resolution     = res_map[resolution_idx];
+
+    st->codecpar->frame_size  = 1024;
+    st->codecpar->sample_rate = sr_map[sr_idx];
+
+    switch (content_type)
+    {
+    case 0:
+        chan_map = av3a_chan_config_table[ch_idx];
+        av_log(c->fc, AV_LOG_INFO, "Output Stream AVS3P3/Audio Vivid: %d Hz, %s(%d channels), s%d, %d kbps \n",
+            st->codecpar->sample_rate, chan_map.tag, chan_map.channels, resolution, brt_kbps);
+        break;
+    case 1:
+        av_log(c->fc, AV_LOG_INFO, "Output Stream AVS3P3/Audio Vivid: %d Hz, %d objects, s%d, %d kbps \n",
+            st->codecpar->sample_rate, num_objects, resolution, brt_kbps);
+        break;
+    case 2:
+        chan_map = av3a_chan_config_table[ch_idx];
+        av_log(c->fc, AV_LOG_INFO, "Output Stream AVS3P3/Audio Vivid: %d Hz, %s + %d objects (%d channels), s%d, %d kbps \n",
+            st->codecpar->sample_rate, chan_map.tag, num_objects, chan_map.channels + num_objects, resolution, brt_kbps);
+        break;
+    case 3:
+        av_log(c->fc, AV_LOG_INFO, "Output Stream AVS3P3/Audio Vivid: %d Hz, %drd order Ambisonics (%d channels), s%d, %d kbps \n",
+            st->codecpar->sample_rate, hoa_order, (hoa_order + 1) * (hoa_order + 1), resolution, brt_kbps);
+    default:
+        break;
+    }
+    st->codecpar->bit_rate = brt_kbps * 1000;
+
+    return 0;
+}
+
 static int mov_read_ddts(MOVContext *c, AVIOContext *pb, MOVAtom atom)
 {
 #define DDTS_SIZE 20
@@ -10059,6 +10157,7 @@ static const MOVParseTableEntry mov_default_parse_table[] = {
 { MKTAG('d','a','c','3'), mov_read_dac3 }, /* AC-3 info */
 { MKTAG('d','e','c','3'), mov_read_dec3 }, /* EAC-3 info */
 { MKTAG('d','d','t','s'), mov_read_ddts }, /* DTS audio descriptor */
+{ MKTAG('d','c','a','3'), mov_read_dca3 }, /* AVS3 Audio Vivid info */
 { MKTAG('w','i','d','e'), mov_read_wide }, /* place holder */
 { MKTAG('w','f','e','x'), mov_read_wfex },
 { MKTAG('c','m','o','v'), mov_read_cmov },
diff --git a/libavformat/movenc.c b/libavformat/movenc.c
index 367caec..6136d33 100644
--- a/libavformat/movenc.c
+++ b/libavformat/movenc.c
@@ -955,6 +955,20 @@ static int mov_write_dmlp_tag(AVFormatContext *s, AVIOContext *pb, MOVTrack *tra
     return update_size(pb, pos);
 }
 
+static int mov_write_dca3_tag(AVFormatContext *s, AVIOContext *pb, MOVTrack *track)
+{
+    int64_t pos = avio_tell(pb);
+
+    avio_wb32(pb, 0);               /* Size */
+    ffio_wfourcc(pb, "dca3");       /* Type */
+
+    if (track->par->extradata_size >= 7) {
+        avio_write(pb, track->par->extradata, 7);
+    }
+
+    return update_size(pb, pos);
+}
+
 static int mov_write_SA3D_tag(AVFormatContext *s, AVIOContext *pb, MOVTrack *track)
 {
     const AVDictionaryEntry *str = av_dict_get(track->st->metadata, "SA3D", NULL, 0);
@@ -1537,6 +1551,8 @@ static int mov_write_audio_tag(AVFormatContext *s, AVIOContext *pb, MOVMuxContex
         ret = mov_write_dops_tag(s, pb, track);
     else if (track->par->codec_id == AV_CODEC_ID_TRUEHD)
         ret = mov_write_dmlp_tag(s, pb, track);
+    else if (track->par->codec_id == AV_CODEC_ID_AVS3_AUDIO)
+        ret = mov_write_dca3_tag(s, pb, track);
     else if (tag == MOV_MP4_IPCM_TAG || tag == MOV_MP4_FPCM_TAG) {
         if (track->par->sample_rate > UINT16_MAX)
             mov_write_srat_tag(pb, track);
@@ -9349,6 +9365,7 @@ static const AVCodecTag codec_mp4_tags[] = {
     { AV_CODEC_ID_MOV_TEXT,        MKTAG('t', 'x', '3', 'g') },
     { AV_CODEC_ID_BIN_DATA,        MKTAG('g', 'p', 'm', 'd') },
     { AV_CODEC_ID_MPEGH_3D_AUDIO,  MKTAG('m', 'h', 'm', '1') },
+    { AV_CODEC_ID_AVS3_AUDIO,      MKTAG('a', 'v', '3', 'a') },
     { AV_CODEC_ID_TTML,            MOV_MP4_TTML_TAG          },
     { AV_CODEC_ID_TTML,            MOV_ISMV_TTML_TAG         },
     { AV_CODEC_ID_FFV1,            MKTAG('F', 'F', 'V', '1') },
diff --git a/libavformat/mpegts.c b/libavformat/mpegts.c
index a1b6065..21c71c4 100644
--- a/libavformat/mpegts.c
+++ b/libavformat/mpegts.c
@@ -852,6 +852,7 @@ static const StreamType ISO_types[] = {
     { STREAM_TYPE_VIDEO_AVS2,     AVMEDIA_TYPE_VIDEO, AV_CODEC_ID_AVS2       },
     { STREAM_TYPE_VIDEO_AVS3,     AVMEDIA_TYPE_VIDEO, AV_CODEC_ID_AVS3       },
     { STREAM_TYPE_VIDEO_VC1,      AVMEDIA_TYPE_VIDEO, AV_CODEC_ID_VC1        },
+    { STREAM_TYPE_AUDIO_AVS3,     AVMEDIA_TYPE_AUDIO, AV_CODEC_ID_AVS3_AUDIO }, /* AVS3 3D Audio Stream Type */
     { 0 },
 };
 
@@ -909,6 +910,7 @@ static const StreamType REGD_types[] = {
     { MKTAG('I', 'D', '3', ' '), AVMEDIA_TYPE_DATA,  AV_CODEC_ID_TIMED_ID3 },
     { MKTAG('V', 'C', '-', '1'), AVMEDIA_TYPE_VIDEO, AV_CODEC_ID_VC1   },
     { MKTAG('O', 'p', 'u', 's'), AVMEDIA_TYPE_AUDIO, AV_CODEC_ID_OPUS  },
+    { MKTAG('a', 'v', '3', 'a'), AVMEDIA_TYPE_AUDIO, AV_CODEC_ID_AVS3_AUDIO},
     { 0 },
 };
 
diff --git a/libavformat/mpegts.h b/libavformat/mpegts.h
index 58ef248..b935bdd 100644
--- a/libavformat/mpegts.h
+++ b/libavformat/mpegts.h
@@ -151,6 +151,7 @@
 #define STREAM_TYPE_VIDEO_CAVS      0x42
 #define STREAM_TYPE_VIDEO_AVS2      0xd2
 #define STREAM_TYPE_VIDEO_AVS3      0xd4
+#define STREAM_TYPE_AUDIO_AVS3      0xd5
 #define STREAM_TYPE_VIDEO_VC1       0xea
 #define STREAM_TYPE_VIDEO_DIRAC     0xd1
 
diff --git a/libavformat/mpegtsenc.c b/libavformat/mpegtsenc.c
index c7733f6..d5f58db 100644
--- a/libavformat/mpegtsenc.c
+++ b/libavformat/mpegtsenc.c
@@ -430,6 +430,9 @@ static int get_dvb_stream_type(AVFormatContext *s, AVStream *st)
     case AV_CODEC_ID_OPUS:
         stream_type = STREAM_TYPE_PRIVATE_DATA;
         break;
+    case AV_CODEC_ID_AVS3_AUDIO:
+        stream_type = STREAM_TYPE_AUDIO_AVS3;
+        break;
     case AV_CODEC_ID_TIMED_ID3:
         stream_type = STREAM_TYPE_METADATA;
         break;
@@ -686,6 +689,10 @@ static int mpegts_write_pmt(AVFormatContext *s, MpegTSService *service)
                 }
             }
 
+            if (codec_id == AV_CODEC_ID_AVS3_AUDIO) {
+                put_registration_descriptor(&q, MKTAG('a', 'v', '3', 'a'));
+            }
+
             if (language != default_language ||
                 st->disposition & (AV_DISPOSITION_CLEAN_EFFECTS    |
                                    AV_DISPOSITION_HEARING_IMPAIRED |
diff --git a/libavformat/rawenc.c b/libavformat/rawenc.c
index cf298d2..9f1ede6 100644
--- a/libavformat/rawenc.c
+++ b/libavformat/rawenc.c
@@ -148,6 +148,23 @@ const FFOutputFormat ff_avs3_muxer = {
 #endif
 
 
+#if CONFIG_AV3A_MUXER
+const FFOutputFormat ff_av3a_muxer = {
+    .p.name            = "av3a",
+    .p.long_name       = NULL_IF_CONFIG_SMALL("AVS3-P3 3D Audio/Audio Vivid AATF bitstream"),
+    .p.mime_type       = "audio/av3a",
+    .p.extensions      = "av3a",
+    .p.audio_codec     = AV_CODEC_ID_AVS3_AUDIO,
+    .p.video_codec     = AV_CODEC_ID_NONE,
+    .p.subtitle_codec  = AV_CODEC_ID_NONE,
+    .flags_internal    = FF_OFMT_FLAG_MAX_ONE_OF_EACH |
+                         FF_OFMT_FLAG_ONLY_DEFAULT_CODECS,
+    .write_packet      = ff_raw_write_packet,
+    .p.flags           = AVFMT_NOTIMESTAMPS,
+};
+#endif
+
+
 #if CONFIG_CAVSVIDEO_MUXER
 const FFOutputFormat ff_cavsvideo_muxer = {
     .p.name            = "cavsvideo",
